File: src/main/java/org/pyjjs/scheduler/core/api/impl/actors/task/behaviours/ResourceAppearedBehaviour.java
package org.pyjjs.scheduler.core.api.impl.actors.task.behaviours;

import org.pyjjs.scheduler.core.api.impl.actors.resource.supervisor.messages.ResourceAppearedMessage;
import org.pyjjs.scheduler.core.api.impl.actors.task.TaskActorState;
import org.pyjjs.scheduler.core.api.impl.actors.task.messages.IFindAnyPlacementMessage;

public class ResourceAppearedBehaviour extends TaskBehaviour<ResourceAppearedMessage> {

    @Override
    protected void perform(ResourceAppearedMessage message) {
        TaskActorState state = getActorState();
        if(hasNotPlacement(state)) {
            send(message.getResourceRef(), new IFindAnyPlacementMessage(getActorRef(), state.getSource().getResourceCriteria()));
        }
    }

    private boolean hasNotPlacement(TaskActorState taskActorState) {
        return taskActorState.getDiscontent() == null;
    }

    @Override
    protected Class<ResourceAppearedMessage> processMessage() {
        return ResourceAppearedMessage.class;
    }
}

File: src/main/kotlin/org/pyjjs/scheduler/core/api/impl/actors/task/behaviours/ResourceAppearedBehaviour.kt
package org.pyjjs.scheduler.core.api.impl.actors.task.behaviours

import org.pyjjs.scheduler.core.api.impl.actors.common.messages.IFindAnyPlacementMessage
import org.pyjjs.scheduler.core.api.impl.actors.common.messages.ResourceAppearedMessage
import org.pyjjs.scheduler.core.api.impl.actors.task.TaskActorState
import kotlin.reflect.KClass

class ResourceAppearedBehaviour : TaskBehaviour<ResourceAppearedMessage>() {

    override fun perform(message: ResourceAppearedMessage) {
        if (hasNotPlacement()) {
            send(message.resourceRef, IFindAnyPlacementMessage(
                    actorRef,
                    actorState.source.descriptor,
                    actorState.source.resourceCriteria))
            actorState.status = TaskActorState.Status.WAIT_OFFER
        }
    }

    private fun hasNotPlacement(): Boolean {
        return TaskActorState.Status.PLACED != actorState.status
    }

    override fun processMessage(): KClass<ResourceAppearedMessage> {
        return ResourceAppearedMessage::class
    }
}
